llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h

```diff
@@ -460,6 +460,16 @@ class DWARFDebugNames : public DWARFAcceleratorTable {
     /// Returns the Offset of the DIE within the containing CU or TU.
     std::optional<uint64_t> getDIEUnitOffset() const;
 
+    /// Returns true if this Entry has information about its parent DIE (i.e. if
+    /// it has an IDX_parent attribute)
+    bool hasParentInformation() const;
+
+    /// Returns the Entry corresponding to the parent of the DIE represented by
+    /// `this` Entry. If the parent is not in the table, nullopt is returned.
+    /// Precondition: hasParentInformation() == true.
+    /// An error is returned for ill-formed tables.
+    Expected<std::optional<DWARFDebugNames::Entry>> getParentDIEEntry() const;
```

Comment:
Do we want expected _and_ optional? Seems a bit much to double unwrap each time we call this?

Reply:
(Drive-by) can you return some kind of no-data error for the parent-not-present case?

Reply:
Mmmm let me look into that, but I am not aware of other instances in LLVM where we use run-time type checking on the Error class to make decisions 🤔

Reply:
Since we have the boolean that tell us if the entry has a parent index case, can we remove the need for an error if we get a `std::nullopt` back but `Entry::hasParentInformation() == true`? I do agree this is nice for `llvm-dwarfdump` to get an error back and display it. If we just use the error somewhere, then it is worth adding.

Reply:
Another idea would be to have this function return an expected pair, where there is a bool indicating if there is parent info + Entry. Something like:
```
Expected<std::pair<bool, DWARFDebugNames::Entry>> getParentDIEEntry() const;
```
I am not a fan of std::pair in general though, so maybe this isn't any better.

Reply:
I think you reinvented std::optional there ;-)

Reply:
> (Drive-by) can you return some kind of no-data error for the parent-not-present case?

I like the clean separation of errors that need to be logged/reported and unsuccessful lookups the current interface provides, as clumsy as it looks.

Reply:
Fair

Reply:
So, given this discussion, I would like to propose we move forward with the interface as is.
As much as it _looks_ off, I did put a lot of thought into this and given that our brainstorm session here did not yield a better result, it seems appropriate.
This interface catches the reader's attention -- as it should -- because there are nuances on the return value that are only explained by reading the docs.